Bulk road edit

dabreegster opened the pull request:
# Problem

Today, we have a tool to bulk edit many lanes at once. You select a bunch of roads, and then just blindly search/replace from one lane type to another:
![screencast](https://user-images.githubusercontent.com/1664407/118566698-396e3c80-b729-11eb-9177-7b201dae3765.gif)

This has pretty limited use, from my knowledge. And now that we have a more detailed road-wide editor, we can do better.

# Simple demo

![screencast](https://user-images.githubusercontent.com/1664407/118566887-9538c580-b729-11eb-9699-73b8ee10065e.gif)
With this new tool, I can first modify one road, changing the parking lane on either side to a bike lane. Then I hit "edit multiple", and the UI automatically finds **all** roads in the map that look similar to the road I originally edited -- same number and type of lanes from before my edit. By default, it starts me with just the subset of these roads that have the same name as my original road. But other roads throughout the map that look similar are highlighted, and I can opt them in as well by clicking to enable/disable. Then I can apply the changes to all roads. That's it!

One complication to address in the future is direction of the road. Internally we match based on "forwards" and "backwards", but when the underlying OSM way splits, this can lead to unexpected results. Here's a hypothetical scenario:
![Screenshot from 2021-05-17 16-07-53](https://user-images.githubusercontent.com/1664407/118567182-2f990900-b72a-11eb-9269-bef15f368a3f.png)
Say the user edits only one side of the left road, then applies the change to the right. If the two roads happen to point towards each other (aka, their internal direction is flipped relative to each other), then the corresponding edit would modify the other side of the road. This isn't what the user would want. We can be more clever by matching up sides of the road based on north/south/east/west-ness, I think. I haven't happened to find an example of this problem yet, but when I do, I'll work on it.

# Yuwen's idea

We discussed this after a user study a few weeks back, but I don't think there are any mocks in Figma yet. The idea was to start with the current road editor, but add a big (+) icon to each end of the road:
![Screenshot from 2021-05-17 16-11-44](https://user-images.githubusercontent.com/1664407/118567489-bfd74e00-b72a-11eb-8d13-4a60e7111f78.png)
When you click it, we select either the next road segment in that direction, or as many segments in that direction until we run out of matching candidates (based on road name and lane counts). There'd be a way to individually deselect some of the segments in the middle of this expansion.

I think this concept would be much more discoverable and intuitive, and we can work towards it from the current PR by adding the (+) icons and making `SelectMultiple` start in just one direction. But I'll leave that as future work, because I think the current thing moves us forward significantly.

# Real use case: Poundbury

For the actdev Poundbury demo, this is what I originally had to do:
![screencast](https://user-images.githubusercontent.com/1664407/118567762-3ecc8680-b72b-11eb-978d-11c6ba34550b.gif)

Imagine that for a few more minutes, making the entire road one-way each direction.

Now it gets much easier:
![screencast](https://user-images.githubusercontent.com/1664407/118567875-6facbb80-b72b-11eb-8cc9-c5df98c60d47.gif)

dabreegster commented on game/src/edit/multiple_roads.rs:
```diff
@@ -0,0 +1,186 @@
+                Line("Click a road segment to select/deselect it."),
+            ])
+            .into_widget(ctx),
+            ColorLegend::row(ctx, Color::RED, "road you've changed"),
```
Wording and colors probably need some tuning

dabreegster commented:
Yuwen suggested not even showing road segments that have the same configuration, but not the same name. Agreed -- simpler to just focus on some approximation of the same "road". And suggested using 3 styles of one color:
![image](https://user-images.githubusercontent.com/1664407/118575064-74c53700-b73a-11eb-8c43-3d27f91cee2a.png)
We used to support some hatching in the shader, but got rid of it in https://github.com/a-b-street/abstreet/commit/07bf9725d424307e0356696b72e7ce6a281214ed#diff-097bf78c0d017f39f80a5bf2e084a38c53cbf29da8fc7c5d46503ec69cd3dd79. It may be worth bringing back, but I remember it had flickering problems anyway because of floating point issues converting back to map-space. So... for now, trying 3 variations of #204AA1:

![Screenshot from 2021-05-17 18-03-37](https://user-images.githubusercontent.com/1664407/118574911-4d6e6a00-b73a-11eb-85ac-4862867e48f2.png)

michaelkirk commented:
Do you want me to review this as is - or were you planning to first make some changes from Yuwen's feedback?

> We used to support some hatching in the shader, but got rid of it in 07bf972#diff-097bf78c0d017f39f80a5bf2e084a38c53cbf29da8fc7c5d46503ec69cd3dd79. It may be worth bringing back

Is this something you could add a sprite and use `Fill::ColoredTexture` for?

dabreegster commented:
> Do you want me to review this as is - or were you planning to first make some changes from Yuwen's feedback?

As it is, please. I prefer many small PRs, but if that's a high review burden, I can batch them more.

> Is this something you could add a sprite and use Fill::ColoredTexture for?

Huh, good idea. I'll try that out. I thought I'd have to mess around with spritesheet packing for some reason, but the png is laid out in quite a straightforward way.

michaelkirk reviewed (approved):
Some UX feedback from my test run of the feature:

- it wasn't clear to me which segments were selected
- it wasn't clear to me that all selectable segments had been selected automatically (Much later I saw the text that said all had been selected, but I'd toggled some lanes already by the time I saw it)
- I didn't realize that I was also editing disparate far away segments (same name and same lane config, but like a mile away)
- I was surprised that "edit multiple" wasn't available until after I'd applied my edits to a single segment. The error pop-up explained it well enough though, so it wasn't a big deal.

## suggestions

- Show the number of lanes currently selected somewhere. Maybe instead of "All road segments with the same original lane configuration...", it could be "{} road segments with the same original lane configuration...", though that text is kind of long and far away from where I'm clicking, so I might not notice it.

- I think Yuwen's suggestions about hatching/texture vs color will help with the confusion around selection

- maybe only automatically select contiguous segments? There's admittedly a downside in that in same cases you will be requiring people to do more work.

Anyway, I don't consider any of of this feedback blocking, just stuff to consider as you iterate.

LGTM!

michaelkirk commented:
Also re: texture

it'd be tricky to use texture to do a stripe-hatching exactly like shown in the screenshot... e.g. consider a road whose direction is parallel to the direction of the hatching, it might look really weird.

But it should work out OK if you use some kind of cross hatching.

dabreegster commented:
Thanks for the detailed UX feedback! I'll address some of that in this PR and ping you again, but I'm not sure about breaking up continuous chunks of road. It does make me imagine some kind of painting controls where you drag the change from the original road in each direction and have it snap to candidate roads...

> it'd be tricky to use texture to do a stripe-hatching exactly like shown in the screenshot... e.g. consider a road whose direction is parallel to the direction of the hatching, it might look really weird.

Actually any kind of geometric texture is going to take at least a little bit of finessing, so that it's a [tesselation](https://en.wikipedia.org/wiki/Tessellation). I mocked up some quick diagonal cross-hatching patterns and quickly realized there's a bit of math involved to make the lines match up when tiled:
![Screenshot from 2021-05-19 09-03-49](https://user-images.githubusercontent.com/1664407/118846337-79483780-b881-11eb-8efc-d350437f7f2d.png)
![spritesheet](https://user-images.githubusercontent.com/1664407/118846391-81a07280-b881-11eb-8841-e4d8db8daf6f.png)
This particular pattern tiles, but the colors and line size are way off. I'll keep messing around; I think plain horizontal and vertical lines will be much easier than diagonals...

dabreegster commented:
PTAL. I'll leave hatching for a followup

> it wasn't clear to me that all selectable segments had been selected automatically

Changed to start with nothing selected, and followed your suggestion to print the number of selected roads. Also added a "select all/none" button, to hopefully help out different use cases people might have.

> I was surprised that "edit multiple" wasn't available until after I'd applied my edits to a single segment

We're now disabling the button if the road is unchanged

michaelkirk commented:
>> I was surprised that "edit multiple" wasn't available until after I'd applied my edits to a single segment
>
> We're now disabling the button if the road is unchanged

I should have been clearer: My expectation from that button is that I would be able to select multiple roads, and then edit them all, rather than editing one road, and then using "edit multiple" to apply those same edits.

For example, I think this is how the traffic signal editor "edit multiple" behaves, right?

I can't think of a quick fix and I'm not sure if this particular change makes things better or worse. It's always unfortunate to push a button that says "you can't do that", but at least the old way allowed you to discover **why** this was the case. 

I'm fine with either way for now - just wanted to clarify my initial comment.

michaelkirk commented:
> Changed to start with nothing selected, and followed your suggestion to print the number of selected roads. Also added a "select all/none" button, to hopefully help out different use cases people might have.

This ended up great! Thanks for working on that.

LGTM overall

dabreegster commented:
Thanks for reviewing!

> It's always unfortunate to push a button that says "you can't do that", but at least the old way allowed you to discover why this was the case.

Thanks for clarifying. Yeah, ideally we would allow selecting and editing in either order. This might be less complicated code than the signal editor's `BundleEdits`, but still too ambitious to attempt now.

Depending on any feedback other people give and whether we attempt implementing the other order, maybe we can go back to a popup to explain the limitation.
